src/System.IO.Pipelines/System/IO/Pipelines/ReadCursor.cs

```diff
@@ -155,9 +192,19 @@ private ReadCursor SeekMultiSegment(long bytes, ReadCursor end, bool checkEndRea
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal void BoundsCheck(ReadCursor newCursor)
```

Comment:
I don't think it makes sense to have a BoundCheck that only verifies the newCursor comes after.

Every time we bounds check, we should be able to come up with both a lower and upper bound. If we are bounds checking at all, it means we don't really trust the input. Having an API that allows you to specify one boundary encourages doing only a partial bounds check.